### ISSUE TITLE
Empty Output When Using CLI

### ISSUE THREAD

**Issue by JulianPitney** [abria/TeraStitcher#74]
**Background:** Terastitcher (GUI version) has been working well for us for a while now. It's use has become a routine part of our microscopy pipelines. Since we're doing it so often I decided I should automate it with the CLI. I've gone through all the documentation multiple times and tried to use it to write a script that does exactly what we've been doing with the GUI. This script is shown below. 

```
:: Step 1 (Import)
set volumePath="C:\Projects\terastitch_script\test"
set ref1="x"
set ref2="y"
set ref3="z"
set vxl1="1.43"
set vxl2="1.43"
set vxl3="5.0"
set imin_regex="000000.tif"
set imin_plugin="tiff3D"
set volin_plugin="TiledXY|3Dseries"
terastitcher --import --volin=%volumePath% --ref1=%ref1% --ref2=%ref2% --ref3=%ref3% --vxl1=%vxl1% --vxl2=%vxl2%^
 --vxl3=%vxl3% --volin_plugin=%volin_plugin% --imin_regex=%imin_regex% --imin_plugin=%imin_plugin%

:: Step 2 (Align)
set USECUDA_X_NCC=1
set projin=%volumePath%\xml_import.xml
set /a oH=245
set /a oV=205
set /a sH=99
set /a sV=99
set /a sD=99
terastitcher --displcompute --projin=%projin% --oH=%oH% --oV=%oV% --sH=%sH% --sV=%sV% --sD=%sD%

:: Step 3 (Project)
set projin=%volumePath%\xml_displcomp.xml
terastitcher --displproj --projin=%projin%

:: Step 4 (Threshold)
set projin=%volumePath%\xml_displproj.xml
set threshold="0.7"
terastitcher --displthres --projin=%projin% --threshold=%threshold%

:: Step 5 (Place)
set projin=%volumePath%\xml_displthres.xml
terastitcher --placetiles --projin=%projin%

:: Step 6 (Merge)
set projin=%volumePath%\xml_merging.xml
set volout="C:\Projects\terastitch_script\result"
set volout_plugin="TiledXY|3Dseries"
set imout_plugin="tiff3D"
set /a resolutions=1
terastitcher --merge --projin=%projin% --volout=%volout% --volout_plugin=%volout_plugin% --resolutions=%resolutions% --imout_plugin=%imout_plugin%
```

**Problem:** At first this script seems to work (All the correct directories and files are created and all the steps appear to run successfully). However, when I open the stitched/output tiff stack, the stack is completely empty (all 0's/all black). 

I've been troubleshooting this all night now but still can't nail down the problem. So far I've tried:

- Triple-checking that I'm following all the examples for steps 1-6 on the wiki,
- Triple-checking all the parameters being given to each step. (Reread the documentation for each parameter multiple times).
- Trying multiple different tiled scans from our microscopes that we've successfully stitched using the GUI version. 
- Desperately trying random parameter values (CUDA on or off, various output resolutions, various output image plugins, etc) 

Please let me know if you have any ideas. I'll update this post as I make progress. Thanks in advance.

**Comment by iannellog** [comment]
I have no time now to examine your problem in detail.
In the meantime, to help me could you please:
- perform the stitching with the GUI and tell me if it works
- perform just the import and align steps on the same image with both the GUI and the CLI script and send me the xml files that are generated (two xml for each execution mode)

**Comment by JulianPitney** [comment]
No problem. I appreciate you taking the time to respond at all. In regard to your questions:

- I've successfully stitched all my scans using the GUI
- [(https://drive.google.com/drive/folders/1ytLF537F7jQEvjkJnU0kGVZhg87fHkrm?usp=sharing)]

**Comment by iannellog** [comment]
I had a look to your xmls and I do not find any evident problems with them. The only difference seems to be that the default value of the --subvoldim option is 100 in the GUI version and 200 in the CLI version. It should not make any difference for your problem, but you can try to use the same value just to check.

One other information I would like to check just to be sure is the version of TeraStitcher you are using. Please run terastitcher --version and send it to me.

Send me also the xml_merging.xml files of both versions for checking also them.

Finally, try to run the following command (without using the script):

terastitcher --merge --projin="C:\Projects\terastitch_script\test\xml_merging.xml" --volout="C:\Projects\terastitch_script\result" --volout_plugin="TiledXY|3Dseries" --resolutions=1 --imout_plugin="tiff3D"

Run it using the xml_merging.xml file generated by both the GUI and the CLI, and check if there is any difference.

**Comment by JulianPitney** [comment]
- **terastitcher  version: 1.10.18 (with Qt 5.2.1)**
- Here's a Google Drive link to the xml_merging.xml files generated by CLI and GUI
(https://drive.google.com/drive/folders/1eTIkH-Np-lf8-_XGBvi2TWhDiTksiqX5?usp=sharing)
- I tried running the command you supplied using CLI and GUI xml_merging.xml files as input. I didn't find any interesting differences in the output. 

I ran the same scan I was using for testing through the GUI version again just to rule out a problem with my .tif files. The GUI version works perfectly. 

I also installed everything from scratch and tried the same script using the same data on a different machine. There was a large change in the file size of the output. On the original machine, the output is generally about ~50MB. On the new machine, the output was ~500MB. Both were still empty/black however. I should note that the file size of the output when using the GUI version is ~1.9GB. (Everything works correctly in the GUI case).

**Comment by iannellog** [comment]
Sorry for the many questions, but I am trying to figure out what is happening.

Let me resume the situation:
- you are using version 1.10.18 for both GUI and CLI versions
- the image is always NOT EMPY when you run the GUI version
- the image is always EMPTY (black) when you run the CLI version
Please confirm this is your case.

Actually the two versions compute different alignments on your image (I do not know why, possibly for the difference in --subvoldim). Again this should not be the reason of your problems, but it is not easy to be sure. If you can, you could do some more experiments:
- try to run the --merge step with CLI version USING THE xml_merging.xml GENERATED BY THE GUI VERSION
- try to run the --displcompute step with the CLI version, BUT SPECIFYING THE OPTION --subvoldim=100, run the other steps and send to me the xml_merging.xml generated this way.

One more thing is that I noticed that you enabled the USECUDA flag in the script. Actually this is weird, since version 1.10.18 does not contain CUDA code. 

I do not make tests with version 1.10.18 since a long time. I will try it just to check if there is some bug I fixed in versions 1.11 and then forgotten.

**Comment by JulianPitney** [comment]
- you are using version 1.10.18 for both GUI and CLI versions
- the image is always NOT EMPY when you run the GUI version
- the image is always EMPTY (black) when you run the CLI version
All 3 of these are true. 

- try to run the --merge step with CLI version USING THE xml_merging.xml GENERATED BY THE GUI VERSION
I did this. Same result. Empty output. 
- try to run the --displcompute step with the CLI version, BUT SPECIFYING THE OPTION --subvoldim=100, run the other steps and send to me the xml_merging.xml generated this way.
I actually tried this after you pointed out that GUI was 100 and CLI was 200. No change in result. This link: 
(https://drive.google.com/drive/folders/1eTIkH-Np-lf8-_XGBvi2TWhDiTksiqX5?usp=sharing) from my earlier post has the xml_merging_GUI.xml and xml_merging_CLI.xml files. They were both generated with subvoldim=100. 

"One more thing is that I noticed that you enabled the USECUDA flag in the script. Actually this is weird, since version 1.10.18 does not contain CUDA code." Yeah I was initially trying to get CUDA running but then I noticed that the windows binaries don't have CUDA built in. I've tried removing that line but it seems to have no affect either way. I'm assuming terastitcher detects that CUDA is not available and defaults to CPU. 

"I do not make tests with version 1.10.18 since a long time. I will try it just to check if there is some bug I fixed in versions 1.11 and then forgotten." I'm comfortable with C++ debugging, if you want to point me to some .cpp files I can help look.

**Comment by iannellog** [comment]
The CLI version has been used by many people with no problems, so your case is really very weird.
The fact that the alignments computed by the two versions with the same input parameters are different (very different in some cases) prove that there should be something wrong in your executables.
Just to avoid too much tests, could you please let me download just one non empty tile of your dataset?
With one tile I can reproduce a test dataset and check if version 1.10.18 on Windows (are you using Windows, right?) has also problems on my machines.
If you want to do another test, try to download version 1.11.10 which contains only the CLI version and that work fine on my machine stitching one channel, 16 bits, images.
I will write you again as soon as I have news.

**Comment by JulianPitney** [comment]
Here's a Google Drive link to one of our tiles. (https://drive.google.com/file/d/1jBowXoVYjSzmC7rNrstr_6l6AMA7jAP7/view?usp=sharing)

Yes I'm using Windows 10. 

Yeah I'm going to try with 1.11.10. I have no idea how I ended up with the 1.10.18 CLI version. I'll report back after testing 1.11.10

**Comment by iannellog** [comment]
I have just stitched a test dataset using your image (4x4 tiles)
Of course all tiles are equal, hence alignment computation is meaningless. However this does not affect the execution of single steps of the pipeline. The final image is correctly generated.
I used CLI version of release 1.11.10 on Mac OS (however I am sure that it works on all platforms).

**Comment by JulianPitney** [comment]
[SOLVED]

Success! Using 1.11.10 on Windows worked perfectly. 

I also discovered how I ended up using **terastitcher version: 1.10.18 (with Qt 5.2.1)**. I installed the GUI version some months ago and didn't realize it had added the binaries to my PATH. Even though I had added the 1.11.10 binaries to the PATH, it was finding the 1,10,18 version first. Everything is clear now. 

Thank you very much for your help. Greatly appreciated. I'll mark this as solved and close it.
